File: 139-word-break/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        mapper = [False] * (len(s) + 1)
        mapper[len(s)] = True
        for i in range(len(s) - 1, -1, -1):
            for word in wordDict:
                if i + len(word) <= len(s) and s[i:i+len(word)] == word:
                    mapper[i] = mapper[i+len(word)]
                if mapper[i]:
                    break
        return mapper[0]

File: 139-word-break/word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        bool* mapper = new bool[s.size() + 1];
        fill_n(mapper,s.size() + 1, false);
        mapper[s.size()] = true;
        for (int i = s.size() - 1; i >= 0; i--){
            for (string word: wordDict){
                if (i+word.size() <= s.size() && s.substr(i, word.size()) == word)
                    mapper[i] = mapper[i+word.size()];
                if (mapper[i]) break;
            }
        }
    
        return mapper[0];
    }
};